compiler/rustc_resolve/src/late.rs

```diff
@@ -2401,32 +2401,70 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
                         if let Some(expr) = expr {
-                            let constant_item_kind = match item.kind {
-                                ItemKind::Const(..) => ConstantItemKind::Const,
-                                ItemKind::Static(..) => ConstantItemKind::Static,
-                                _ => unreachable!(),
-                            };
                             // We already forbid generic params because of the above item rib,
                             // so it doesn't matter whether this is a trivial constant.
                             this.with_constant_rib(
                                 IsRepeatExpr::No,
                                 ConstantHasGenerics::Yes,
-                                Some((item.ident, constant_item_kind)),
+                                Some((item.ident, ConstantItemKind::Static)),
                                 |this| this.visit_expr(expr),
                             );
                         }
                     });
                 });
             }
 
+            ItemKind::Const(box ast::ConstItem { ref generics, ref ty, ref expr, .. }) => {
+                // FIXME(generic_consts): Is the ordering of ribs correct? Right now, it's
+                // STATIC <- LIFETIME_ANON_REPORT <- GENERICS <- LIFETIME_ELIDED_STATIC {,<- CONSTANT}
+                self.with_static_rib(|this| {
+                    this.with_lifetime_rib(LifetimeRibKind::AnonymousReportError, |this| {
+                        this.with_generic_param_rib(
+                            &generics.params,
+                            RibKind::Item(HasGenericParams::Yes(generics.span)),
+                            LifetimeRibKind::Generics {
+                                binder: item.id,
+                                kind: LifetimeBinderKind::Item,
+                                span: generics.span,
+                            },
+                            |this| {
+                                this.with_lifetime_rib(
+                                    LifetimeRibKind::Elided(LifetimeRes::Static),
+                                    |this| {
+                                        this.visit_generics(generics);
+                                        this.visit_ty(ty);
+                                    },
+                                );
+
+                                this.with_lifetime_rib(
+                                    LifetimeRibKind::Elided(LifetimeRes::Infer),
+                                    |this| {
+                                        let Some(expr) = expr else { return; };
+
+                                        // FIXME(generic_consts): Is this still true?
+                                        // > We already forbid generic params because of the above item rib,
+                                        // > so it doesn't matter whether this is a trivial constant.
```

Comment:
The `Item` rib is outside of the generic params rib, so we do need to decide whether the constant can access generic params. If they cannot, generic consts are not very useful.